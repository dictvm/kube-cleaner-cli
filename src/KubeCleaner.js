'use strict';
const React = require('react');
const importJsx = require('import-jsx');
const { AppContext, render, Box } = require('ink');
const SelectInput = require('ink-select-input').default;
const TextInputWithEnter = importJsx('./TextInputWithEnter');
const ClusterRelationships = importJsx('./ClusterRelationships');

const { updateConfig, writeConfig, loadConfig } = require('./utils');

const KubeCleaner = () => {
  const appContext = React.useContext(AppContext);

  const [config] = React.useState(() => loadConfig());

  const [cluster, setCluster] = React.useState(null);
  const [confirmationText, setConfirmationText] = React.useState('');
  const [deletionConfirmed, setDeletionConfirmed] = React.useState(false);

  const relatedContexts =
    cluster === null
      ? []
      : config.contexts.filter(
          context => context.context.cluster === cluster.name,
        );

  const relatedUsers =
    cluster === null
      ? []
      : config.users.filter(user => user.name.startsWith(cluster.name));

  const clusterOptions = config.clusters.map(cluster => ({
    label: cluster.name,
    value: cluster.name,
    cluster: cluster,
  }));

  React.useEffect(() => {
    if (deletionConfirmed === true) {
      const updatedConfig = updateConfig(
        config,
        cluster,
        relatedContexts,
        relatedUsers,
      );
      writeConfig(updatedConfig);
      appContext.exit();
    }
  }, [deletionConfirmed]);

  return (
    <Box>
      {cluster == null ? (
        <SelectInput
          items={clusterOptions}
          onSelect={item => {
            setCluster(item.cluster);
          }}
        />
      ) : (
        <Box flexDirection="column">
          <ClusterRelationships
            cluster={cluster}
            contexts={relatedContexts}
            users={relatedUsers}
          />
          <Box marginTop={1}>
            Confirm deletion y/n:{' '}
            <TextInputWithEnter
              value={confirmationText}
              onChange={query => {
                setConfirmationText(query);
              }}
              onSubmit={() => {
                if (confirmationText === 'y') {
                  setDeletionConfirmed(true);
                }
                switch (confirmationText) {
                  case 'y':
                    return setDeletionConfirmed(true);
                  case 'n':
                    setDeletionConfirmed(false);
                    setConfirmationText('');
                    return setCluster(null);
                  default:
                    setDeletionConfirmed(null);
                    setConfirmationText('');
                    return;
                }
              }}
            />
          </Box>
          {deletionConfirmed === true ? (
            <Box>Cluster {cluster.name} has been deleted.</Box>
          ) : null}
          {deletionConfirmed === null ? <Box>Invalid input.</Box> : null}
        </Box>
      )}
    </Box>
  );
};

render(<KubeCleaner />);
