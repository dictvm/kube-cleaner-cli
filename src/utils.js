const yml = require('js-yaml');
const fs = require('fs');
const homedir = require('os').homedir();
const path = require('path');

const kubeConfigPath = path.join(homedir, '.kube/config');

module.exports.updateConfig = (kubeConfig, cluster, contexts, users) => {
  const config = JSON.parse(JSON.stringify(kubeConfig));
  const contextNames = contexts.map(context => context.name);
  const userNames = users.map(user => user.name);

  config.clusters = config.clusters.filter(c => c.name !== cluster.name);
  config.contexts = config.contexts.filter(context => {
    return !contextNames.includes(context.name);
  });
  config.users = config.users.filter(user => {
    return !userNames.includes(user.name);
  });

  return config;
};

module.exports.loadConfig = () => {
  const kubeContent = fs.readFileSync(kubeConfigPath, 'utf8');
  return yml.safeLoad(kubeContent);
};

module.exports.writeConfig = config => {
  const yamlConfig = yml.safeDump(config);
  fs.writeFileSync(kubeConfigPath, yamlConfig, 'utf8');
};
