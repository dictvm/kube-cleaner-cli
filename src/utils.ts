import yml from 'js-yaml';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { Cluster, Context, User, Config } from './types';

const kubeConfigPath = path.join(os.homedir(), '.kube', 'config');

export const updateConfig = (
  kubeConfig: Config,
  cluster: Cluster,
  contexts: Context[],
  users: User[],
): Config => {
  const config: Config = JSON.parse(JSON.stringify(kubeConfig));
  const contextNames = contexts.map((context: Context) => context.name);
  const userNames = users.map(user => user.name);

  config.clusters = config.clusters.filter(c => c.name !== cluster.name);
  config.contexts = config.contexts.filter(
    context => !contextNames.includes(context.name),
  );
  if (config['current-context'] === cluster.name) {
    delete config['current-context'];
  }
  config.users = config.users.filter(user => !userNames.includes(user.name));

  return config;
};

export const loadConfig = (): Config => {
  const kubeContent = fs.readFileSync(kubeConfigPath, 'utf8');
  return yml.safeLoad(kubeContent);
};

export const writeConfig = (config: Config) => {
  const yamlConfig = yml.safeDump(config);
  fs.writeFileSync(kubeConfigPath, yamlConfig, 'utf8');
};
