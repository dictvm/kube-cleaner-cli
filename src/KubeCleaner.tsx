import React, { useContext, useState, useEffect } from 'react';
import { AppContext, render, Box } from 'ink';
import SelectInput from 'ink-select-input';
import TextInputWithEnter from './TextInputWithEnter';
import ClusterRelationships from './ClusterRelationships';

import { updateConfig, writeConfig, loadConfig } from './utils';
import { Cluster } from './types';

const KubeCleaner = () => {
  const appContext = useContext(AppContext);

  const [config] = useState(() => loadConfig());

  const [cluster, setCluster] = useState<null | Cluster>(null);
  const [confirmationText, setConfirmationText] = useState('');
  const [deletionConfirmed, setDeletionConfirmed] = useState<boolean | null>(
    false,
  );

  useEffect(() => {
    if (deletionConfirmed === true && cluster !== null) {
      const updatedConfig = updateConfig(
        config,
        cluster,
        relatedContexts,
        relatedUsers,
      );
      writeConfig(updatedConfig);
      appContext.exit();
    }
  }, [deletionConfirmed]);

  const relatedContexts =
    cluster === null
      ? []
      : config.contexts.filter(
          context => context.context.cluster === cluster.name,
        );

  const relatedUsers =
    cluster === null
      ? []
      : config.users.filter(user => user.name.startsWith(cluster.name));

  const clusterOptions = config.clusters.map(cluster => ({
    label: cluster.name,
    value: cluster.name,
    cluster: cluster,
  }));

  if (clusterOptions.length === 0) {
    return <Box>No clusters found in the config</Box>;
  }

  return (
    <Box>
      {cluster == null ? (
        <SelectInput
          items={clusterOptions}
          onSelect={item => {
            setCluster(item.cluster);
          }}
        />
      ) : (
        <Box flexDirection="column">
          <ClusterRelationships
            cluster={cluster}
            contexts={relatedContexts}
            users={relatedUsers}
          />
          <Box marginTop={1}>
            Confirm deletion y/n:{' '}
            <TextInputWithEnter
              value={confirmationText}
              onChange={query => {
                setConfirmationText(query);
              }}
              onSubmit={() => {
                if (confirmationText === 'y') {
                  setDeletionConfirmed(true);
                }
                switch (confirmationText) {
                  case 'y':
                    return setDeletionConfirmed(true);
                  case 'n':
                    setDeletionConfirmed(false);
                    setConfirmationText('');
                    return setCluster(null);
                  default:
                    setDeletionConfirmed(null);
                    setConfirmationText('');
                    return;
                }
              }}
            />
          </Box>
          {deletionConfirmed === true ? (
            <Box>Cluster {cluster.name} has been deleted.</Box>
          ) : null}
          {deletionConfirmed === null ? <Box>Invalid input.</Box> : null}
        </Box>
      )}
    </Box>
  );
};

export const start = () => render(<KubeCleaner />);
