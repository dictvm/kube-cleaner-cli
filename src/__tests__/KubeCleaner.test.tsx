import React from 'react';
import yml from 'js-yaml';
import fs from 'fs';
import path from 'path';
import * as utils from '../utils';
import { KubeCleaner } from '../KubeCleaner';
import { render } from 'ink-testing-library';

test('it works', () => {
  jest.spyOn(utils, 'loadConfig').mockImplementation(() => {
    const configPath = path.resolve(__dirname, '__fixtures__', 'config');
    const content = fs.readFileSync(configPath, 'utf8');
    return yml.safeLoad(content);
  });

  const { lastFrame } = render(<KubeCleaner />);
  expect(lastFrame()).toMatchSnapshot();
});
