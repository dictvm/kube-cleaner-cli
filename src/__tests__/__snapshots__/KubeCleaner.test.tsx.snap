// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it works 1`] = `"[34m‚ùØ[39m [39m[34mdocker-for-desktop-cluster[39m[39m"`;
